### PR TITLE
validate-modules: new module: fail if 'type' isn't documented

### PR DESCRIPTION
**Pull request by pilou-**
##### SUMMARY
validate-modules: new module: fail if `type` isn't documented and type in `arg_spec` isn't `str`.

* `type` isn't required in documentation when `choices` is set (different types can be used with `choices`, [see](https://github.com/ansible/ansible/blob/493cf817a920542fc3d791e1bafdc52c05cbc06d/lib/ansible/module_utils/basic.py#L1572))

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
validate-modules

##### ADDITIONAL INFORMATION
Related: #56312

Locally tested on top of #51746 (70435d615e6e32ec0c37f80f78ce0d505b0b5e11):

```
$ python test/sanity/validate-modules/validate-modules  --arg-spec lib/ansible/modules/cloud/scaleway/scaleway_lb_facts.py --base-branch upstream/devel
lib/ansible/modules/cloud/scaleway/scaleway_lb_facts.py:0:0: E337 Argument 'api_version' in argument_spec implies type as 'str' but documentation doesn't define type
lib/ansible/modules/cloud/scaleway/scaleway_lb_facts.py:0:0: E337 Argument 'region' in argument_spec implies type as 'str' but documentation doesn't define type
```

### CODE REVIEW

**Comment by pilou-** [comment]
Sanity errors below seems unrelated:
```
18:06 ERROR: Found 2 validate-modules issue(s) which need to be resolved:
18:06 ERROR: lib/ansible/modules/cloud/vmware/vsphere_copy.py:0:0: E309 version_added for new option (host) should be '2.9'. Currently None (0%)
18:06 ERROR: lib/ansible/modules/cloud/vmware/vsphere_copy.py:0:0: E309 version_added for new option (login) should be '2.9'. Currently None (0%)
```

**Comment by mattclay** [comment]
@pilou- Those errors have been resolved. I've restarted the failed tests.

**Comment by felixfontein** [comment]
(I think you have a typo, you mean `type='str'` and not `type='list'` for `choices`.)

About the implications: `suboptions` can also be used with `type='dict'`, and `choices` can also be used with other types (such as `int`).

**Comment by pilou-** [comment]
> About the implications: suboptions can also be used with type='dict', and choices can also be used with other types (such as int).

PR updated: `type` is now required when `suboptions` is set

**Comment by mattclay** [comment]
bot_status

**Comment by pilou-** [comment]
@felixfontein @mattclay does that look ok ?

**Comment by mattclay** [comment]
@pilou- New tests should add ignores for existing failures instead of adding `_is_new_module` checks.

**Comment by felixfontein** [comment]
@pilou- I would also require the docs type to be specified for `choices`. Besides that, LGTM.

**Comment by pilou-** [comment]
> I would also require the docs type to be specified for choices.

You're right, i will update this PR.

**Comment by pilou-** [comment]
@felixfontein docs type is required for choices
@mattclay `ignore.txt` has been updated

File: test/sanity/validate-modules/main.py [hunk 1]
```diff
@@ -1245,16 +1245,31 @@ def _validate_argument_spec(self, docs, spec, kwargs):
...
             # TODO: needs to recursively traverse suboptions
             doc_type = docs.get('options', {}).get(arg, {}).get('type')
-            if 'type' in data:
-                if data['type'] != doc_type and doc_type is not None:
+            if 'type' in data and data['type'] is not None:
+                if doc_type is None:
+                    if not arg.startswith('_'):  # hidden parameter, for example _raw_params
+                        self.reporter.error(
+                            path=self.object_path,
+                            code=336,
```
Comment (mattclay) [inline test/sanity/validate-modules/main.py]: Code 336 is already defined: https://docs.ansible.com/ansible/devel/dev_guide/testing_validate-modules.html

Both new codes should be added to the docs.

Comment (pilou-) [reply test/sanity/validate-modules/main.py]: Updated:
- E337/E338 used instead of E336/E337
- new codes added to the doc

**Comment by pilou-** [comment]
@mattclay please, could you take another look ?

**Review by mattclay — APPROVED** [review]

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
Before merging, CI should be run again. Due to recently merged module renames, it could be that some entries in ignore.txt need to be added / removed.

**Comment by mattclay** [comment]
CI restarted. However, there's also a conflict in the ignore file to resolve.

**Comment by pilou-** [comment]
@mattclay sivel won't be available the next week or two, could someone else review ?

**Comment by mattclay** [comment]
@pilou- I restarted the tests to see what new ignores are needed. Add those and I'll go ahead and merge, so the PR doesn't need to be updated again.

**Comment by pilou-** [comment]
@mattclay PR updated.

**Comment by felixfontein** [comment]
While trying to remove the docker modules from the ignore list, I found a potential bug in the validation. For `docker_network`, it yields one error:
```
ERROR: lib/ansible/modules/cloud/docker/docker_network.py:0:0: E337 Argument 'network_name' in argument_spec defines type as 'str' but documentation doesn't define type
```
But `network_name` is an alias for `name`, which has its type documented. The problem is apparently that the documentation says the option name is `name` and its alias is `network_name`, while `argument_spec` says the option is `network_name` with alias `name`. See https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/docker/docker_network.py#L23-L29 and https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/docker/docker_network.py#L634

The main question for me is now: is it allowed to switch option/alias order between argument_spec and docs? If yes, there's a bug in this sanity test. If no, another sanity check would be needed.

**Pull request merged**